cli/src/alluxio.org/cli/processes/common.go

```diff
@@ -12,113 +12,150 @@
 package processes
 
 import (
-	"bufio"
 	"fmt"
 	"io"
+	"io/ioutil"
 	"os"
-	"path"
+	"path/filepath"
 	"strings"
 	"time"
 
+	"github.com/palantir/stacktrace"
 	"golang.org/x/crypto/ssh"
 
 	"alluxio.org/cli/env"
 	"alluxio.org/log"
 )
 
-func getMasters() ([]string, error) {
-	mastersDir := path.Join(env.Env.EnvVar.GetString(env.ConfAlluxioConfDir.EnvVar), "masters")
-	mastersFile, err := os.Open(mastersDir)
-	if err != nil {
-		log.Logger.Errorf("Error reading worker hostnames at %s", mastersDir)
-		return nil, err
+var cliPath = filepath.Join(env.Env.EnvVar.GetString(env.ConfAlluxioHome.EnvVar), "bin", "alluxio")
+var privateKeySigner ssh.Signer
+var masterList []string
+var workerList []string
+
+// for each master, create a client and run
+// TODO: now start master one by one, need to do them in parallel
+func runCommand(command string, onMasters bool, onWorkers bool) error {
+	// get list of masters and workers
+	var nodes []string
+	if onMasters {
+		if len(masterList) == 0 {
+			if err := getNodes(true); err != nil {
+				log.Logger.Fatalf("Cannot get masters, error: %s", err)
+			}
+		}
+		nodes = append(nodes, masterList...)
+	}
+	if onWorkers {
+		if len(workerList) == 0 {
+			if err := getNodes(false); err != nil {
+				log.Logger.Fatalf("Cannot get workers, error: %s", err)
+			}
+		}
+		nodes = append(nodes, workerList...)
+	}
+
+	// get public key if nones
+	if privateKeySigner == nil {
```

Comment:
also need lock and prefer to use sync.Once to fetch the private key. since the private key is needed for all multi-process commands, it seems reasonable to always fetch it once at the beginning of the command and then pass it to the runCommand functions